storage/series_test.go

```diff
@@ -119,3 +123,185 @@ func TestChunkSeriesSetToSeriesSet(t *testing.T) {
+				// Convert all histograms to float histograms.
+				for i := range test.samples {
+					test.samples[i].(*sample).fh = test.samples[i].H().ToFloat()
+					test.samples[i].(*sample).h = nil
+				}
+				testHistogramsSeriesToChunks(t, test)
+			})
+		})
+	}
+	t.Run("mixed_histograms", func(t *testing.T) {
+		mixedTests := []histogramTest{{
+			name: "histogram and float histogram encoded to two chunks",
+			lbs:  labels.FromStrings("__name__", "up", "instance", "localhost:8080"),
+			samples: []tsdbutil.Sample{
+				&sample{t: 1, h: h1},
+				&sample{t: 2, fh: h2.ToFloat()},
+			},
+			expectedChunks: 2,
+		}, {
+			name: "histogram and float histogram encoded to two chunks",
+			lbs:  labels.FromStrings("__name__", "up", "instance", "localhost:8080"),
+			samples: []tsdbutil.Sample{
+				&sample{t: 1, fh: h1.ToFloat()},
+				&sample{t: 2, h: h2},
+			},
+			expectedChunks: 2,
+		}}
+		for _, test := range mixedTests {
+			testHistogramsSeriesToChunks(t, test)
+		}
+	})
+}
+
+func testHistogramsSeriesToChunks(t *testing.T, test histogramTest) {
+	series := NewListSeries(test.lbs, test.samples)
+	encoder := NewSeriesToChunkEncoder(series)
+	require.EqualValues(t, test.lbs, encoder.Labels())
+
+	chks, err := ExpandChunks(encoder.Iterator(nil))
+	require.NoError(t, err)
+	require.Equal(t, test.expectedChunks, len(chks))
+
+	// Decode all encoded samples and assert they are equal to the original ones.
+	encodedSamples := expandHistogramSamples(chks)
+	require.Equal(t, len(test.samples), len(encodedSamples))
+	for i, encodedSample := range encodedSamples {
+		var testSample *histogram.FloatHistogram
+		if test.samples[i].H() != nil {
+			testSample = test.samples[i].H().ToFloat()
+		} else {
+			testSample = test.samples[i].FH()
+		}
+
+		if value.IsStaleNaN(testSample.Sum) {
+			require.True(t, value.IsStaleNaN(encodedSample.Sum))
+			continue
+		}
+		require.True(t, encodedSample.Compact(0).Equals(testSample))
+	}
+
+	// If a counter reset hint is expected, it can only be found in the second chunk.
+	// Otherwise, we assert an unknown counter reset hint in all chunks.
+	if test.expectedCounterReset {
+		require.Equal(t, chunkenc.UnknownCounterReset, getCounterResetHint(chks[0]))
+		require.Equal(t, chunkenc.CounterReset, getCounterResetHint(chks[1]))
+	} else {
+		for _, chk := range chks {
+			require.Equal(t, chunkenc.UnknownCounterReset, getCounterResetHint(chk))
+		}
+	}
+}
+
+func expandHistogramSamples(chunks []chunks.Meta) []*histogram.FloatHistogram {
+	if len(chunks) == 0 {
+		return nil
+	}
+
+	return append(
+		expandHistogramChunk(chunks[0].Chunk.Iterator(nil)),
+		expandHistogramSamples(chunks[1:])...,
+	)
+}
+
+func expandHistogramChunk(it chunkenc.Iterator) []*histogram.FloatHistogram {
+	floatHistograms := make([]*histogram.FloatHistogram, 0)
+	for it.Next() != chunkenc.ValNone {
+		_, fh := it.AtFloatHistogram()
+		floatHistograms = append(floatHistograms, fh)
+	}
+	return floatHistograms
+}
```

Comment:
1. I recommend keeping it simple (to read) and merge this two functions into one without recursion.
2. It is better to test the samples as is - i.e. don't turn integer histogram to float histogram, rather test it as integer histogram.